### PR TITLE
[KeyVault] adding msrestazure credential authentication support

### PR DESCRIPTION
**Pull request by schaabs**
Updating key vault client to support both KeyVaultAuthentication and msrestazure Credential object for authentication.

### CODE REVIEW

File: azure-keyvault/azure/keyvault/custom/key_vault_client.py [hunk 1]
```diff
@@ -6,12 +6,43 @@
...
+
+        :param credentials: Credentials needed for the client to connect to Azure.
+        :type credentials: :mod:`A msrestazure Credentials
+         object<msrestazure.azure_active_directory>` or :mod:`A KeyVaultAuthentication
+         object<key_vault_authentication>` 
+        """
+        self._inner_creds = None
+
+        # if the supplied credentials instance is not derived from KeyVaultAuthBase but is an AADTokenCredentials instance
+        if not isinstance(credentials, KeyVaultAuthBase) and isinstance(credentials, AADTokenCredentials):
```
Comment (lmazuel) [inline azure-keyvault/azure/keyvault/custom/key_vault_client.py]: `AADMixin`, not `AADTokenCredentials`. This one is deprecated and not in the hierarchy of `ServicePrincipalCredentials`

File: azure-keyvault/azure/keyvault/custom/key_vault_client.py [hunk 2]
```diff
@@ -6,12 +6,43 @@
...
+
+            # create a callback which authenticates with the supplied credentials instance
+            self._inner_creds = credentials
+
+            def auth_callback(server, resource, scope):
+                self._inner_creds.resource = resource
+                token = self._inner_creds.token
```
Comment (lmazuel) [inline azure-keyvault/azure/keyvault/custom/key_vault_client.py]: `token` is just a property and does not call AD to get the token, so you don't do the call with the new resource.
Add this before to do the call :
```python
self._inner_creds.set_token()
```

**Comment by schaabs** [comment]
@lmazuel I've updated the code based off your feedback and our conversation.   I believe this scheme will work better as it allows KeyVaultAuthentication to handle refresh_session.  I'm still doing some testing but I've updated the PR so you can take a look.

File: azure-keyvault/azure/keyvault/custom/key_vault_authentication.py [hunk 3]
```diff
@@ -140,18 +140,51 @@ def auth_callack(server, resource, scope):
...
         return session
+
+    def refresh_session(self):
+        """Return updated session if token has expired, attempts to
+        refresh using refresh token.
+
+        :rtype: requests.Session.
+        """
+        if self._credentials:
+            self._credentials.refresh_session()
```
Comment (lmazuel) [inline azure-keyvault/azure/keyvault/custom/key_vault_authentication.py]: miss a return

File: azure-keyvault/azure/keyvault/custom/key_vault_authentication.py [hunk 4]
```diff
@@ -140,18 +141,47 @@ def auth_callack(server, resource, scope):
...
         :param authorization_callback: A callback used to provide authentication credentials to the key vault data service.  
         This callback should take three str arguments: authorization uri, resource, and scope, and return 
         a tuple of (token type, access token).
+        :param credentials:: Credentials needed for the client to connect to Azure.
+        :type credentials: :mod:`A msrestazure Credentials
+         object<msrestazure.azure_active_directory>`
         """
-        super(KeyVaultAuthentication, self).__init__()
+        if not authorization_callback and not credentials:
+            raise ValueError("Either parameter 'authorization_callback' or parameter 'credentials' must not be None.")
```
Comment (RandalliLama) [inline azure-keyvault/azure/keyvault/custom/key_vault_authentication.py]: This seems like a double negative.  How about:

Either parameter 'authorization_callback' or parameter 'credentials' must be specified.

File: azure-keyvault/azure/keyvault/custom/key_vault_authentication.py [hunk 5]
```diff
@@ -140,18 +141,47 @@ def auth_callack(server, resource, scope):
...
+        # super(KeyVaultAuthentication, self).__init__()
+
+        self._credentials = credentials
+
+        if not authorization_callback:
+            def auth_callback(server, resource, scope):
+                if self._credentials.resource != resource:
+                    self._credentials.resource = resource
+                    token = self._credentials.set_token()
+                token = self._credentials.token
```
Comment (RandalliLama) [inline azure-keyvault/azure/keyvault/custom/key_vault_authentication.py]: I don't understand 165 and 166.  Won't 166 just undo what was done at 165, or is there some Python magic going on here?

**Pull request merged**